propolis-server-lib/tests/integration_tests.rs

```diff
@@ -0,0 +1,235 @@
+        "disk".to_string(),
+        toml::value::Value::String(
+            TEST_IMAGE.path().as_path().to_str().unwrap().to_string(),
+        ),
+    );
+    block_options.insert(
+        "pci-path".to_string(),
+        toml::value::Value::String("0.4.0".to_string()),
+    );
+    let block = Device {
+        driver: "pci-virtio-block".to_string(),
+        options: block_options,
+    };
+    let mut devices = BTreeMap::new();
+    devices.insert("block0".to_string(), block);
+
+    let config = Config::new(TEST_BOOTROM.path(), devices);
+    let context = server::Context::new(config);
+
+    let config_dropshot = ConfigDropshot {
+        bind_address: "127.0.0.1:0".parse().unwrap(),
+        ..Default::default()
+    };
+    HttpServerStarter::new(&config_dropshot, server::api(), context, &log)
+        .unwrap()
+        .start()
+}
+
+// Create a new instance.
+//
+// NOTE: Many of these values are placeholders, and can be replaced
+// when we have "real" integration of UUID-based images / bootroms.
+fn create_ensure_request(
+    id: Uuid,
+) -> propolis_client::api::InstanceEnsureRequest {
+    propolis_client::api::InstanceEnsureRequest {
+        properties: propolis_client::api::InstanceProperties {
+            id,
+            name: "test-instance".to_string(),
+            description: "it's a test instance".to_string(),
+            image_id: Uuid::new_v4(),
+            bootrom_id: Uuid::new_v4(),
+            memory: 256,
+            vcpus: 2,
+        },
+    }
+}
+
+// Utility wrapper for monitoring an instance's state until it reaches a
+// requested target.
+async fn wait_until_state(
+    client: &Client,
+    id: Uuid,
+    state: InstanceState,
+) -> Result<()> {
+    let mut gen = 0;
+    loop {
+        let monitor_response = client.instance_state_monitor(id, gen).await?;
+        if monitor_response.gen < gen {
+            bail!(
+                "Gen should be increasing: (requested {}, saw {})",
+                gen,
+                monitor_response.gen
+            );
+        }
+        eprintln!(
+            "Monitor response: {:?} (waiting for {:?})",
+            monitor_response, state
+        );
+        if monitor_response.state == state {
+            return Ok(());
+        }
+        gen = monitor_response.gen + 1;
+    }
+}
+
+#[tokio::test]
+async fn test_uninitialized_server() {
+    let log = initialize_log();
+    let server = initialize_server(log.clone()).await;
+
+    let client = Client::new(server.local_addr(), log);
+    assert!(matches!(
+        client.instance_get(Uuid::nil()).await,
+        Err(ClientError::Status(500))
+    ));
+
+    server.close().await.unwrap();
+}
+
+#[tokio::test]
+async fn test_ensure_instance_put_running() {
+    let log = initialize_log();
+    let server = initialize_server(log.clone()).await;
+
+    let client = Client::new(server.local_addr(), log);
+
+    let id = Uuid::from_str("0000002a-000c-0005-0c03-0938362b0809").unwrap();
```

Comment:
These UUIDs are arbitrary but consistent - this means that if a test is failing repeatedly, we won't be generating tons of distinct VMs on a machine.

Reply:
Sounds good for until we have a reason for it to be otherwise